src/compute/src/sink/persist_sink.rs

```diff
@@ -1209,3 +1218,114 @@ where
     let token = Rc::new(shutdown_button.press_on_drop());
     (output_stream, token)
 }
+
+/// This is for REFRESH options on materialized views. It adds an operator that rounds up the
+/// timestamps of data and frontiers to the time of the next refresh. See
+/// `doc/developer/design/20231027_refresh_every_mv.md`.
+///
+/// Note that this currently only works with 1-dim timestamps. (This is not an issue for WMR,
+/// because iteration numbers should disappear by the time the data gets to the Persist sink.)
+fn apply_refresh<G>(
+    coll: Collection<G, Result<Row, DataflowError>, Diff>,
+    refresh_schedule: RefreshSchedule,
+) -> Collection<G, Result<Row, DataflowError>, Diff>
+where
+    G: Scope<Timestamp = Timestamp>,
+{
+    // We need to disconnect the reachability graph and manage capabilities manually, because we'd like to round up
+    // frontiers as well as data: as soon as our input frontier passes a refresh time, we'll round it up to the next
+    // refresh time.
+    let mut builder = OperatorBuilder::new("apply_refresh".to_string(), coll.scope());
+    let (mut output_buf, output_stream) = builder.new_output();
+    let mut input = builder.new_input_connection(&coll.inner, Pipeline, vec![Antichain::new()]);
+    builder.build(move |capabilities| {
+        // This capability directly controls this operator's output frontier (because we have disconnected the input
+        // above). We wrap it in an Option so we can drop it to advance to the empty output frontier when the last
+        // refresh is done. (We must be careful that we only ever emit output updates at times that are at or beyond
+        // this capability.)
+        let mut capability = Some(capabilities.into_element());
+        let mut buffer = Vec::new();
+        move |frontiers| {
+            let mut output_handle_core = output_buf.activate();
+            let mut output_buf = ConsolidateBuffer::new(&mut output_handle_core, 0);
+            input.for_each(|input_cap, data| {
+                // Note that we can't use `input_cap` to get an output session because we might have advanced our output
+                // frontier already beyond the frontier of this capability.
+
+                // `capability` will be None if we are past the last refresh. We have made sure to not receive any
+                // data that is after the last refresh by setting the `until` of the dataflow to the last refresh.
+                let capability = capability
+                    .as_mut()
+                    .expect("should have a capability if we received data");
```

Comment:
How certain are we about this part? It sounds reasonable, on first glance. But we only provide the `until` to index and source import operators, and we don't have anything in place that prevents other upstream operators from inventing updates from times in the future, right?

I would feel better if we handled this gracefully by soft-asserting and returning instead.

Reply:
One thing that we briefly checked is that the rendering of temporal filters seems to be also mindful of the `until`: if it were to delay some addition or retraction to after the `until`, then it drops those instead.

Reply:
We had a meeting with @teskje and @antiguru and decided to handle this and similar cases with soft panics in this operator.